### PR TITLE
Enable embedding sources to Windows PDBs

### PR DESCRIPTION
**Pull request by tmat**
Finishes implementation of https://github.com/dotnet/roslyn/issues/12625 for Windows PDBs.

The feature was only completed for Portable PDBs due to bugs in Microsoft.DiaSymReader.Native that are now fixed.

The first couple of commits change the PDB validation to treat order of data as significant. Previously we used AssertXml helper that ignores order.

Resolves https://github.com/dotnet/roslyn/issues/13707.

### CODE REVIEW

File: build/Targets/Packages.props [hunk 1]
```diff
@@ -34,10 +34,10 @@
     <MicrosoftCSharpVersion>4.3.0</MicrosoftCSharpVersion>
     <MicrosoftDiagnosticsRuntimeVersion>0.8.31-beta</MicrosoftDiagnosticsRuntimeVersion>
     <MicrosoftDiagnosticsTracingTraceEventVersion>1.0.35</MicrosoftDiagnosticsTracingTraceEventVersion>
-    <MicrosoftDiaSymReaderVersion>1.1.0</MicrosoftDiaSymReaderVersion>
-    <MicrosoftDiaSymReaderConverterXmlVersion>1.0.0-beta1-61708-01</MicrosoftDiaSymReaderConverterXmlVersion>
-    <MicrosoftDiaSymReaderNativeVersion>1.6.0-beta2-25304</MicrosoftDiaSymReaderNativeVersion>
-    <MicrosoftDiaSymReaderPortablePdbVersion>1.3.0</MicrosoftDiaSymReaderPortablePdbVersion>
+    <MicrosoftDiaSymReaderVersion>1.2.0-beta1-62008-01</MicrosoftDiaSymReaderVersion>
+    <MicrosoftDiaSymReaderConverterXmlVersion>1.1.0-beta1-62008-01</MicrosoftDiaSymReaderConverterXmlVersion>
+    <MicrosoftDiaSymReaderNativeVersion>1.7.0-private-25604</MicrosoftDiaSymReaderNativeVersion>
```
Comment (tmat) [inline build/Targets/Packages.props]: I'll replace this with official bits once they are available.

**Comment by tmat** [comment]
@richlander FYI

**Comment by tmat** [comment]
@jaredpar The unit test run failed with


```

```

**Comment by tmat** [comment]
test windows_debug_unit32_prtest please

**Comment by jaredpar** [comment]
That's bizzare. It's not failing with that anywhere else and I don't see anything in yourchange that would cause that

File: src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs [hunk 2]
```diff
@@ -1535,6 +1565,22 @@ public unsafe void EmbedSourceLink(Stream stream)
             }
         }
 
+        /// <summary>
+        /// Write document entries for any embedded text document that does not yet have an entry.
+        /// </summary>
+        /// <remarks>
+        /// This is done after serializing method debug info to ensure that we embed all requested
+        /// text even if there are no correspodning sequence points.
```
Comment (nguerrera) [inline src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs]: Typo: corresponding. (This was me, sorry.)

**Comment by tannergooding** [comment]
@tmat, will this require any coordination when inserted? @brettfo for possible FYI.

**Comment by tmat** [comment]
@tannergooding No insertion coordination is needed.

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 3]
```diff
@@ -1981,14 +1979,32 @@ class Program {
...
+            {
+                case "/debug:embedded":
+                    ValidateEmbeddedSources_Portable(expectedEmbeddedMap, dir, isEmbeddedPdb: true);
+                    break;
+                case "/debug:portable":
+                    ValidateEmbeddedSources_Portable(expectedEmbeddedMap, dir, isEmbeddedPdb: false);
+                    break;
+                case "/debug:full":
+                    ValidateEmbeddedSources_Windows(expectedEmbeddedMap, dir);
+                    break;
```
Comment (cston) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs]: Is `"/debug:full"` tested?

File: src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb [hunk 4]
```diff
@@ -2884,9 +2899,30 @@ print Goodbye, World"
...
+
+                    Dim sourceStr = Encoding.UTF8.GetString(sourceBlob.Array, sourceBlob.Offset, sourceBlob.Count)
+
+                    Assert.Equal(expectedEmbeddedMap(docPath), sourceStr)
+                    Assert.True(expectedEmbeddedMap.Remove(docPath))
+                Next
+            Finally
+                symReader?.Dispose()
+            End Try
         End Sub
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb]: Can these helper methods be shared between C# and VB, perhaps in `PdbTestUtilities`?

**Comment by cston** [comment]
LGTM

File: build/Targets/Packages.props (hunk 1 above)
Comment (tmat) [inline build/Targets/Packages.props]: Turns out VCTools are not being inserted into VS 15.5 yet. There won't be any official builds until next week. As long as we don't insert the private bits into VS (https://github.com/dotnet/roslyn/pull/21420 removes DSRN from insertion) it is ok to use them.

Comment (jaredpar) [inline build/Targets/Packages.props]: Is this the version of the native PDB writer that supports full determinism?

Comment (tmat) [inline build/Targets/Packages.props]: Yes, that change is in the branch I built from.

Comment (jaredpar) [inline build/Targets/Packages.props]: Can you pull in the powershell change from this PR? 

https://github.com/dotnet/roslyn/pull/20527/files

That should pass now and keep us clean.

Comment (tmat) [inline build/Targets/Packages.props]: Sure.

Comment (tmat) [inline build/Targets/Packages.props]: Actually, let me do that in a separate PR. Determinism is not really related to this change.

Comment (jaredpar) [inline build/Targets/Packages.props]: Gotcha. Don't worry about a second PR then. I'll just rebase mine against these changes once they're merged.

Comment (tmat) [inline build/Targets/Packages.props]: OK. Have you seen my email to YongKang about PDBs not being deterministic when writing twice within the same process? Might affect compiler server.

Comment (jaredpar) [inline build/Targets/Packages.props]: Yeah I saw that. It didn't seem to affect our tests for some reason. Maybe just got lucky. But definitely need to get that fixed before we declare it "done".

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Can I have a second compiler sign off please?
